Further work

braver opened the pull request:
Implements the most urgent pull requests and adds general polish.

devallabharath committed:
Update iTerm2-v3.sh

solved multiple popups on mac with iTerm2

jpe90 committed:
Add example for kitty on OSX

braver committed:
clean up menu items and use split view

braver committed:
clean up and update readme

windows line endings 🤨
shorter and clearer

braver committed:
Merge branch 'master' of github.com:jpe90/sublime_terminal into improvements

braver committed:
Merge branch 'master' of github.com:devallabharath/sublime_terminal into improvements

braver committed:
sublime always runs python3 these days

braver committed:
tell pyright we're developing for all systems

braver committed:
stop worrying about python 2

braver committed:
spin some logic out to functions, pass flake8

braver committed:
fall back to ~ if no directory found

ie. when the current file context has not been saved to disk yet
do this while flashing a status bar message that this happened

closes https://github.com/wbond/sublime_terminal/pull/211

reasoning for not introducing more logic or settings:
if you create a file at ~/project/some_dir/new_file.txt
and open the terminal from that file's context,
it's just as unexpected that the terminal opens at ~/project instead of
~/project/some_dir, as it would be if the terminal opens at ~:
it's not what the user asked for,
but it's still better than throwing an error

this also ensures get_path always returns a path, simplifying stuff

braver committed:
remove the commands if they're not applicable

ie. there is no file or there is no project root
closes https://github.com/wbond/sublime_terminal/pull/120

braver committed:
commands are not interactive: should not have ...

braver committed:
cleanup stuff we don't use and old messages/logs

braver committed:
add lint job

braver committed:
specify noqa

braver committed:
elif -> if since we're returning

braver committed:
readme readability

FichteFoll commented on Terminal.py:
```diff
@@ -22,14 +23,80 @@ def get_setting(key, default=None):
     os_specific_settings = {}
     if os.name == 'nt':
-        os_specific_settings = sublime.load_settings('Terminal (Windows).sublime-settings')
+        os_specific_settings = sublime.load_settings(
+            'Terminal (Windows).sublime-settings')
```
Just my personal opinion, but this is pretty ugly. I suggest using a higher maximum line width, I like 120.
I also like having the closing paren on a new line with the same indentation as the opening one, but YMMV.

braver replied:
Yeah, it feels weird. Personally I've gotten used to it as a typical "python thing". But you're right, let's just bump the line length to 120.  I do that all the time, 80 just always gets you in a pickle with long strings anyway and although I like the ambition of 80 (I always have a ruler there) it's just always just a hassle.

FichteFoll commented on Terminal.py:
```diff
@@ -142,27 +142,28 @@ def get(terminal_key):
+        sublime.status_message('Terminal: opening at home directory')
+        return os.path.expanduser('~')
+
+    def run_terminal(self, dir, terminal, parameters):
```
This shadows a built-in (you're probably aware of this but it's usually bad practice).

braver replied:
No I wasn't aware actually, I was just moving stuff around. It can have any name I guess, I'll rename it.

braver replied:
Or do you mean the key bindings? That’s a known issue. I kinda want to remove them, and just document how users can set up their own, but it’s a widely used package and it’s been this way forever. So that’s going to cause some friction.

braver replied:
Eh, what the hey, no time like the present. Let's drop those bindings.

FichteFoll replied:
No, I just meant the `dir` parameter. The key bindings is indeed a known issue with the package, though.

braver replied:
oh, right, sure, I can change that 👍🏻

FichteFoll commented on Terminal.py:
```diff
@@ -188,6 +188,13 @@ def run_terminal(self, dir, terminal, parameters):
 
 
 class OpenTerminalCommand(sublime_plugin.WindowCommand, TerminalCommand):
+    def is_visible(self):
+        # remove the command if the view doesn't have a path to open at
+        # taking is_visible over is_enabled to remove it from the context menu,
+        # instead of simply disabling the entry
+        view = self.window.active_view()
+        return bool(view and view.file_name())
```
I'm thinking, opening in a default folder (e.g. `~`, via the last commit) *might* be preferrable, but it should be reflected in the command's/menu label when this happens. This can be accomplished by overriding `description`.

That said, I personally just press super+enter when I need a terminal regardless of the location & window I have currently active and maybe others do, too.

braver replied:
My thinking was that opening "a terminal" is really an OS thing, not a ST thing. The context menu should have things that make sense for the context, and simply opening a terminal at `~` is not "contextual". The command palette might have a separate "Terminal: open at ~" command in it though, I could see how that might be convenient.

It's also very much an edge case. This package only has "no solution" if you're editing a newly create file in a window with no directory opened. When does that ever happen?

braver replied:
Besides, the current behaviour is to throw an error message and do nothing useful. So we could consider new features of course, but this change just prevents users from getting into the situation where the package say "I don't know what to do, good luck" in the first place :D

FichteFoll replied:
Sure, this is fine as-is.

FichteFoll commented on messages.json:
```diff
@@ -1,6 +0,0 @@
-	"1.7.0": "messages/1.7.0.txt",
-	"1.8.0": "messages/1.8.0.txt",
-	"1.13.0": "messages/1.13.0.txt"
-}
```
Just making sure that these changelogs are recorded in some place (e.g. github releases) and not entirely "lost" afterwards, excluding the commit history.

braver replied:
Personally I'm never really so attached to these things. The changelog has nothing that the git history, or GitHub, doesn't have, and doesn't add any depth in terms of explanations or anything. And git will keep them if anyone wants to read it. I also don't believe humanity needs a record of every bug fix in a little bit of code like this. Not to make a value judgement on anyone else's work though. So let me know if you want these kept in a text file somewhere.

FichteFoll replied:
It's fine. This isn't my decision, I just wanted to confirm it's something you thought about.

braver committed:
120 line length: the luxury of large screens

braver committed:
rename run_terminal

braver committed:
fix missing command in sidebar context menu

in that case paths are passed, but there might not be a view

braver committed:
fix menu and palette entries for preferences

the palette still had split default/user entries
the main settings entry was broken

braver committed:
breaking: stop shipping with key bindings

they mask built-ins
fixes https://github.com/wbond/sublime_terminal/issues/216
fixes https://github.com/wbond/sublime_terminal/issues/17

braver committed:
add messaging around the key bindings change

FichteFoll commented on Default.sublime-commands:
```diff
@@ -1,64 +1,27 @@
         "args": {
-            "file": "${packages}/User/Terminal.sublime-settings"
+            "base_file": "${packages}/Terminal/Terminal.sublime-settings",
+            "user_file": "${packages}/User/Terminal ($platform).sublime-settings",
```
OS-specific settings files are not supported for arbitrary settings files afaik. This only works for the main preferences file. Unless you added support for it in code, of course.

braver replied:
This has been here forever and I kind of assumed it did something. OS specific user settings indeed don't work for user settings, I checked (for Open in Default Application). I'll have a look, see if this is any kind of useful or can simply be dropped.

braver replied:
`get_setting` does read the user's OS specific settings and uses them to override the normal settings. So yeah, it's implemented here. I guess that's useful if you're syncing ST settings across machines with different operating systems.

braver committed:
don't shadow dir

FichteFoll reviewed (approved):
LGTM. All comments have been resolved.

FichteFoll commented:
For context, these have been integrated into the fork at https://github.com/SublimeText/Terminal.
